Mathlib/CategoryTheory/SmallObject/Iteration/UniqueHom.lean

```diff
@@ -0,0 +1,144 @@
+        iter₂.isoZero.inv ≫ eqToHom (by congr; symm; simpa using hi)
+      naturality := fun ⟨i, hi⟩ ⟨k, hk⟩ φ => by
+        obtain rfl : i = ⊥ := by simpa using hi
+        obtain rfl : k = ⊥ := by simpa using hk
+        obtain rfl : φ = 𝟙 _ := rfl
+        simp }
+  natTrans_app_succ i hi := by simp at hi
+
+section
+
+variable {i : J} {iter₁ iter₂ : Iteration ε (Order.succ i)}
+  (hi : ¬IsMax i) (φ : iter₁.trunc (SuccOrder.le_succ i) ⟶ iter₂.trunc (SuccOrder.le_succ i))
+
+/-- Auxiliary definition for `mkOfSucc`. -/
+noncomputable def mkOfSuccNatTransApp (k : J) (hk : k ≤ Order.succ i) :
+    iter₁.F.obj ⟨k, hk⟩ ⟶ iter₂.F.obj ⟨k, hk⟩ :=
+  if hk' : k = Order.succ i then
+    eqToHom (by subst hk'; rfl) ≫ (iter₁.isoSucc i (Order.lt_succ_of_not_isMax hi)).hom ≫
+      whiskerRight (φ.natTrans.app ⟨i, by simp⟩) _ ≫
+      (iter₂.isoSucc i (Order.lt_succ_of_not_isMax hi)).inv ≫
+      eqToHom (by subst hk'; rfl)
+  else
+    φ.natTrans.app ⟨k, Order.le_of_lt_succ (by
+      obtain hk | rfl := hk.lt_or_eq
+      · exact hk
+      · tauto)⟩
+
+lemma mkOfSuccNatTransApp_eq_of_le (k : J) (hk : k ≤ i) :
+    mkOfSuccNatTransApp hi φ k (hk.trans (Order.le_succ i)) =
+      φ.natTrans.app ⟨k, hk⟩ :=
+  dif_neg (by rintro rfl; simpa using lt_of_le_of_lt hk (Order.lt_succ_of_not_isMax hi))
+
+@[simp]
+lemma mkOfSuccNatTransApp_succ_eq :
+    mkOfSuccNatTransApp hi φ (Order.succ i) (by rfl) =
+      (iter₁.isoSucc i (Order.lt_succ_of_not_isMax hi)).hom ≫
+        whiskerRight (φ.natTrans.app ⟨i, by simp⟩) _ ≫
+        (iter₂.isoSucc i (Order.lt_succ_of_not_isMax hi)).inv := by
+  dsimp [mkOfSuccNatTransApp]
+  rw [dif_pos rfl, comp_id, id_comp]
+
+/-- Auxiliary definition for `mkOfSucc`. -/
+@[simps]
+noncomputable def mkOfSuccNatTrans :
+    iter₁.F ⟶ iter₂.F where
+  app := fun ⟨k, hk⟩ => mkOfSuccNatTransApp hi φ k hk
+  naturality := fun ⟨k₁, hk₁⟩ ⟨k₂, hk₂⟩ f => by
+    dsimp
+    have hk : k₁ ≤ k₂ := leOfHom f
+    obtain h₂ | rfl := hk₂.lt_or_eq
+    · replace h₂ : k₂ ≤ i := Order.le_of_lt_succ h₂
+      rw [mkOfSuccNatTransApp_eq_of_le hi φ k₂ h₂,
+        mkOfSuccNatTransApp_eq_of_le hi φ k₁ (hk.trans h₂)]
+      exact natTrans_naturality φ k₁ k₂ hk h₂
+    · obtain h₁ | rfl := hk.lt_or_eq
+      · have h₂ : k₁ ≤ i := Order.le_of_lt_succ h₁
+        let f₁ : (⟨k₁, hk⟩ : { l | l ≤ Order.succ i}) ⟶
+          ⟨i, Order.le_succ i⟩ := homOfLE h₂
+        let f₂ : (⟨i, Order.le_succ i⟩ : Set.Iic (Order.succ i)) ⟶
+          ⟨Order.succ i, by simp⟩ := homOfLE (Order.le_succ i)
+        obtain rfl : f = f₁ ≫ f₂ := rfl
+        rw [Functor.map_comp, Functor.map_comp, assoc,
+          mkOfSuccNatTransApp_eq_of_le hi φ k₁ h₂]
+        erw [← natTrans_naturality_assoc φ k₁ i h₂ (by rfl)]
+        rw [mkOfSuccNatTransApp_succ_eq]
+        dsimp
+        change _ ≫ iter₁.mapSucc i (Order.lt_succ_of_not_isMax hi) ≫ _ =
+          _ ≫ _ ≫ iter₂.mapSucc i (Order.lt_succ_of_not_isMax hi)
```

Comment:
I think it would be nicer to have this as
```suggestion
        have ha : iter₁.F.map f₂ = iter₁.mapSucc i (Order.lt_succ_of_not_isMax hi) := rfl
        have hb : iter₂.F.map f₂ = iter₂.mapSucc i (Order.lt_succ_of_not_isMax hi) := rfl
        rw [ha, hb]
```
because then it's less magical what is happening. I'm not familiar with all of the definitions involved, but perhaps it could even be possible to extract a lemma?